File: src/main/java/br/com/fiap/siase/service/PedidoCompraService.java
package br.com.fiap.siase.service;

import br.com.fiap.siase.dto.request.PedidoCompraRequest;
import br.com.fiap.siase.dto.response.PedidoCompraResponse;
import br.com.fiap.siase.exception.BusinessException;
import br.com.fiap.siase.exception.ResourceNotFoundException;
import br.com.fiap.siase.model.PedidoCompra;
import br.com.fiap.siase.model.enums.StatusPedidoCompra;
import br.com.fiap.siase.repository.PecaRepository;
import br.com.fiap.siase.repository.PedidoCompraRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.UUID;

@Service
@RequiredArgsConstructor
public class PedidoCompraService {

    private final PedidoCompraRepository repository;
    private final PecaRepository pecaRepository;

    @Transactional
    public PedidoCompraResponse criar(PedidoCompraRequest request) {
        var peca = pecaRepository.findById(request.pecaId())
                .orElseThrow(() -> new ResourceNotFoundException("Peça não encontrada: " + request.pecaId()));

        var pedido = new PedidoCompra();
        pedido.setPeca(peca);
        pedido.setQuantidadeSolicitada(request.quantidadeSolicitada());
        pedido.setObservacoes(request.observacoes());

        return PedidoCompraResponse.from(repository.save(pedido));
    }

    @Transactional(readOnly = true)
    public List<PedidoCompraResponse> listar() {
        return repository.findAll().stream().map(PedidoCompraResponse::from).toList();
    }

    @Transactional(readOnly = true)
    public List<PedidoCompraResponse> listarPorStatus(StatusPedidoCompra status) {
        return repository.findByStatus(status).stream().map(PedidoCompraResponse::from).toList();
    }

    @Transactional(readOnly = true)
    public PedidoCompraResponse buscarPorId(UUID id) {
        return PedidoCompraResponse.from(findOrThrow(id));
    }

    @Transactional
    public PedidoCompraResponse aprovar(UUID id) {
        var pedido = findOrThrow(id);
        try {
            pedido.aprovar();
        } catch (IllegalStateException e) {
            throw new BusinessException(e.getMessage());
        }
        return PedidoCompraResponse.from(repository.save(pedido));
    }

    @Transactional
    public PedidoCompraResponse receber(UUID id, int quantidadeRecebida) {
        var pedido = findOrThrow(id);

        if (quantidadeRecebida <= 0) {
            throw new BusinessException("Quantidade recebida deve ser maior que zero.");
        }

        try {
            pedido.receber(quantidadeRecebida);
        } catch (IllegalStateException e) {
            throw new BusinessException(e.getMessage());
        }

        pecaRepository.save(pedido.getPeca());
        return PedidoCompraResponse.from(repository.save(pedido));
    }

    @Transactional
    public PedidoCompraResponse cancelar(UUID id) {
        var pedido = findOrThrow(id);
        try {
            pedido.cancelar();
        } catch (IllegalStateException e) {
            throw new BusinessException(e.getMessage());
        }
        return PedidoCompraResponse.from(repository.save(pedido));
    }

    private PedidoCompra findOrThrow(UUID id) {
        return repository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Pedido de compra não encontrado: " + id));
    }
}

File: frontend/src/app/core/services/pedido-compra.service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';
import { PedidoCompraRequest, PedidoCompraResponse, StatusPedidoCompra } from '../models/pedido-compra.model';
import { environment } from '../../../environments/environment';

@Injectable({ providedIn: 'root' })
export class PedidoCompraService {
  private readonly http = inject(HttpClient);
  private readonly base = `${environment.apiUrl}/pedidos-compra`;

  listar(status?: StatusPedidoCompra) {
    const params = status ? new HttpParams().set('status', status) : undefined;
    return this.http.get<PedidoCompraResponse[]>(this.base, { params });
  }

  buscarPorId(id: string) {
    return this.http.get<PedidoCompraResponse>(`${this.base}/${id}`);
  }

  criar(body: PedidoCompraRequest) {
    return this.http.post<PedidoCompraResponse>(this.base, body);
  }

  aprovar(id: string) {
    return this.http.patch<PedidoCompraResponse>(`${this.base}/${id}/aprovar`, null);
  }

  receber(id: string, quantidade: number) {
    return this.http.patch<PedidoCompraResponse>(
      `${this.base}/${id}/receber`,
      null,
      { params: new HttpParams().set('quantidade', quantidade) }
    );
  }

  cancelar(id: string) {
    return this.http.patch<PedidoCompraResponse>(`${this.base}/${id}/cancelar`, null);
  }
}
